include/deal.II/lac/constrained_linear_operator.h

```diff
@@ -0,0 +1,292 @@
+ * of linear equations:
+ * @f[
+ *   (C^T A C + Id_c) x = C^T (b - A\,k)
+ * @f]
+ * with a given (unconstrained) system matrix $A$, right hand side $b$, and
+ * linear constraints $C$ with inhomogeneities $k$.
+ *
+ * A detailed explanation of this approach is given in the @ref constraints
+ * module.
+ *
+ * @author Mauro Bardelloni, Matthias Maier, 2015
+ *
+ * @relates LinearOperator
+ * @ingroup constraints
+ */
+template <typename Range, typename Domain>
+LinearOperator<Range, Domain>
+constrained_linear_operator(const ConstraintMatrix &constraint_matrix,
+                            const LinearOperator<Range, Domain> &linop)
+{
+  const auto C =
+    distribute_constraints_linear_operator(constraint_matrix, linop);
+  const auto Ct = transpose_operator(C);
+  const auto Id_c =
+    project_to_constrained_linear_operator(constraint_matrix, linop);
+  return Ct * linop * C + Id_c;
+}
+
+
+/**
+ * Given a ConstraintMatrix object @p constraint_matrix, a LinearOperator
+ * @p linop and a right-hand side @p right_hand_side, this function creates
+ * a PackagedOperation that stores the following computation:
+ * @code
+ *   Ct * (right_hand_side - linop * k)
+ * @endcode
+ * with
+ * @code
+ *   C = distribute_constraints_linear_operator(constraint_matrix, linop);
+ *   Ct = transpose_operator(C);
+ * @endcode
+ *
+ * This LinearOperator object is used together with
+ * constrained_right_hand_side() to build up the following modified system
+ * of linear equations:
+ * @f[
+ *   (C^T A C + Id_c) x = C^T (b - A\,k)
+ * @f]
+ * with a given (unconstrained) system matrix $A$, right hand side $b$, and
+ * linear constraints $C$ with inhomogeneities $k$.
+ *
+ * A detailed explanation of this approach is given in the @ref constraints
+ * module.
+ *
+ * @author Mauro Bardelloni, Matthias Maier, 2015
+ *
+ * @relates LinearOperator
+ * @ingroup constraints
+ */
+template <typename Range, typename Domain>
+PackagedOperation<Range>
+constrained_right_hand_side(const ConstraintMatrix &constraint_matrix,
+                            const LinearOperator<Range, Domain> &linop,
+                            const Range &right_hand_side)
+{
+  PackagedOperation<Range> return_comp;
+
+  return_comp.reinit_vector = linop.reinit_range_vector;
+
+  return_comp.apply_add =
+    [&constraint_matrix, &linop, &right_hand_side](Range &v)
+  {
+    const auto C =
+      distribute_constraints_linear_operator(constraint_matrix, linop);
+    const auto Ct = transpose_operator(C);
+
+    static GrowingVectorMemory<Domain> vector_memory;
+    Domain *k = vector_memory.alloc();
+    linop.reinit_domain_vector(*k, /*bool fast=*/ false);
+    constraint_matrix.distribute(*k);
+
+    v += Ct * (right_hand_side - linop **k);
+
+    vector_memory.free(k);
+  };
+
+  // lambda capture expressions are a C++14 feature...
+  const auto apply_add = return_comp.apply_add;
+  return_comp.apply = [apply_add](Range &v)
```

Comment:
Oh, I see. I'd say just remove the comment then, it's only understandable if you know where the author originally wanted to go with the code.